internal/namespace/manager.go

```diff
@@ -0,0 +1,143 @@
+	GetConfigWithFallback(ns string, key config.RootKey) string
+}
+
+type namespaceManager struct {
+	ctx      context.Context
+	database database.Plugin
+	nsConfig map[string]config.Section
+}
+
+func NewNamespaceManager(ctx context.Context, di database.Plugin) (Manager, error) {
+	if di == nil {
+		return nil, i18n.NewError(ctx, coremsgs.MsgInitializationNilDepError)
+	}
+	nm := &namespaceManager{
+		ctx:      ctx,
+		database: di,
+		nsConfig: buildNamespaceMap(ctx),
+	}
+	return nm, nil
+}
+
+func buildNamespaceMap(ctx context.Context) map[string]config.Section {
+	conf := namespacePredefined
+	namespaces := make(map[string]config.Section, conf.ArraySize())
+	for i := 0; i < conf.ArraySize(); i++ {
+		nsConfig := conf.ArrayEntry(i)
+		name := nsConfig.GetString(coreconfig.NamespaceName)
+		if name != "" {
+			if _, ok := namespaces[name]; ok {
+				log.L(ctx).Warnf("Duplicate predefined namespace (ignored): %s", name)
+			}
+			namespaces[name] = nsConfig
+		}
+	}
+	return namespaces
+}
+
+func (nm *namespaceManager) Init(ctx context.Context) error {
+	return nm.initNamespaces(ctx)
+}
+
+func (nm *namespaceManager) getPredefinedNamespaces(ctx context.Context) ([]*core.Namespace, error) {
+	defaultNS := config.GetString(coreconfig.NamespacesDefault)
+	namespaces := []*core.Namespace{
+		{
+			Name:        core.LegacySystemNamespace,
+			Type:        core.NamespaceTypeSystem,
+			Description: i18n.Expand(ctx, coremsgs.CoreSystemNSDescription),
+		},
+	}
+	i := 0
+	foundDefault := false
+	for name, nsObject := range nm.nsConfig {
+		if err := core.ValidateFFNameField(ctx, name, fmt.Sprintf("namespaces.predefined[%d].name", i)); err != nil {
+			return nil, err
+		}
+		i++
+		foundDefault = foundDefault || name == defaultNS
+		namespaces = append(namespaces, &core.Namespace{
+			Type:        core.NamespaceTypeLocal,
+			Name:        name,
+			Description: nsObject.GetString("description"),
+		})
+	}
+	if !foundDefault {
+		return nil, i18n.NewError(ctx, coremsgs.MsgDefaultNamespaceNotFound, defaultNS)
+	}
+	return namespaces, nil
+}
+
+func (nm *namespaceManager) initNamespaces(ctx context.Context) error {
+	predefined, err := nm.getPredefinedNamespaces(ctx)
+	if err != nil {
+		return err
+	}
+	for _, newNS := range predefined {
+		ns, err := nm.database.GetNamespace(ctx, newNS.Name)
+		if err != nil {
+			return err
+		}
+		var updated bool
+		if ns == nil {
+			updated = true
+			newNS.ID = fftypes.NewUUID()
+			newNS.Created = fftypes.Now()
+		} else {
+			// Only update if the description has changed, and the one in our DB is locally defined
+			updated = ns.Description != newNS.Description && ns.Type == core.NamespaceTypeLocal
+		}
+		if updated {
+			if err := nm.database.UpsertNamespace(ctx, newNS, true); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (nm *namespaceManager) GetConfigWithFallback(ns string, key config.RootKey) string {
+	if nsConfig, ok := nm.nsConfig[ns]; ok {
```

Comment:
Per other comments, I feel like this isn't top-level on the namespace, rather one level down - on the multi-party system identity specific section.

Reply:
@awrichar - did you have any thoughts on this spelling point?